prototype2/client/src/renderer/opengl/mod.rs

```diff
@@ -124,7 +135,31 @@ impl OpenGlRenderer {
     data.camera_pos = [camera_adj_pos.0, camera_adj_pos.1, camera_adj_pos.2];
     gfx_window_glutin::update_views(window, &mut data.out_color, &mut data.out_depth);
     // encoder.update_constant_buffer(&data.locals, &locals);
-    encoder.draw(&slice, &self.pso, data);
+    encoder.draw(&slice, &self.pso_3d, data);
+  }
+
+  pub fn render_ui(&mut self,
+                   encoder: &mut gfx::Encoder<gfx_device_gl::Resources,
+                                              gfx_device_gl::CommandBuffer>,
+                   window: &mut glutin::Window,
+                   debug_msg: &debug::DebugMessage,
+                   menu_state: &MenuState) {
+
+    let &debug::DebugMessage(ref message) = debug_msg;
+    gfx_window_glutin::update_views(window, &mut self.main_color, &mut self.main_depth);
+    // Add some text 10 pixels down and right from the top left screen corner.
+    self.text_renderer.add(message, // Text to add
+                           [10, 10], // Position
+                           [0.9, 0.9, 0.9, 1.0] /* Text color */);
+
+    if menu_state.is_open() {
+      self.text_renderer.add("Menu is open \n ", // Text to add
+                             [50, 50], // Position
+                             [0.9, 0.9, 0.9, 1.0] /* Text color */);
```

Comment:
Yeah its pretty good. The guts are pretty messy though, and I'm not confident its not making unneccessary allocations, but its definitely sufficient for now.

It supports taking a projection matrix which I think will let me render text in the world. Thats all of  my use cases, so i shouldn't have to unearth my own rendering code.